### PR TITLE
plan, expression: improve merge join with multiple keys.

### PR DESCRIPTION
**Pull request by hanfei1991**
Consider all possible properties for merge join.

### CODE REVIEW

File: plan/dag_plan_test.go [hunk 1]
```diff
@@ -211,10 +211,11 @@ func (s *testPlanSuite) TestDAGPlanBuilderJoin(c *C) {
 			best: "SemiJoinWithAux{TableReader(Table(t))->TableReader(Table(t))}->Projection",
 		},
 		// Test Single Merge Join.
-		{
-			sql:  "select /*+ TIDB_SMJ(t1,t2)*/ * from t t1, t t2 where t1.a = t2.b",
-			best: "MergeJoin{TableReader(Table(t))->TableReader(Table(t))->Sort}(t1.a,t2.b)",
-		},
+		// Merge Join will no longer enforce a sort.
```
Comment (winoros) [inline plan/dag_plan_test.go]: This test can be removed?

File: plan/new_physical_plan_builder.go [hunk 2]
```diff
@@ -280,21 +278,89 @@ func (p *LogicalJoin) generatePhysicalPlans() []PhysicalPlan {
...
+			otherConds = append(otherConds, eqCond)
+		} else {
+			eqConds = append(eqConds, eqCond)
+		}
+	}
+	return eqConds, otherConds
+}
+
+func (p *LogicalJoin) getMergeJoin() []PhysicalPlan {
+	//	log.Warnf("")
```
Comment (winoros) [inline plan/new_physical_plan_builder.go]: remove the log.

File: plan/new_physical_plan_builder.go [hunk 3]
```diff
@@ -282,22 +280,89 @@ func (p *LogicalJoin) generatePhysicalPlans() []PhysicalPlan {
...
+	tmpSchema := expression.NewSchema(cols2...)
+	permutation := make([]int, 0, len(cols1))
+	for i, col1 := range cols1 {
+		offset := tmpSchema.ColumnIndex(col1)
+		if offset == -1 {
+			return permutation, cols1[:i]
+		}
+		permutation = append(permutation, offset)
+	}
+	return permutation, cols1
```
Comment (lamxTyler) [inline plan/new_physical_plan_builder.go]: Why return the `cols1` unchanged?

Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: If all column can match cols1, we return cols1. If part of columns can match cols1, we return part of cols1 in line 289

File: plan/new_physical_plan_builder.go [hunk 4]
```diff
@@ -282,22 +280,89 @@ func (p *LogicalJoin) generatePhysicalPlans() []PhysicalPlan {
 	}
 }
 
-func (p *LogicalJoin) preferUseMergeJoin() bool {
-	return p.preferMergeJoin && len(p.EqualConditions) == 1
+func getPermutation(cols1, cols2 []*expression.Column) ([]int, []*expression.Column) {
```
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: Nope, different logic

**Comment by lamxTyler** [comment]
LGTM

**Pull request merged**